src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3243,6 +3369,20 @@ public IEnumerator<TElement> GetEnumerator()
                 for (int i = 0; i < buffer.count; i++) yield return buffer.items[map[i]];
             }
         }
+        
+        public TElement[] ToArray()
+        {
+            Buffer<TElement> buffer = new Buffer<TElement>(source);
+            int count = buffer.count;
+            TElement[] array = new TElement[count];
+            if (count > 0)
+            {
+                int[] map = GetEnumerableSorter(null).Sort(buffer.items, count);
```

Comment:
I'll have them both go through a `private int[] SortedMap(Buffer<TElement> buffer)` method, that should be small enough to be inlined in most cases.

Otherwise I'd worry that maybe some compiler wouldn't have the sorter only as a local within `MoveNext()` of the produced iterator? While the current coding comes close to guaranteeing it becomes a field, it firmly guarantees such a field gets cleared again in the same `MoveNext()` call that sets it.
If some compilers create a field for the unnamed sorter, then that could have a GC impact, since the sorter can be quite large (one array full of number-of-source-elements keys for each level of sorting) that wouldn't be great.
MSBuild seems fine with it, but I'd want to be sure.
Going through a separate method removes that risk.